provers/risc0/driver/src/bonsai.rs

```diff
@@ -263,6 +265,30 @@ pub async fn cancel_proof(uuid: String) -> anyhow::Result<()> {
     Ok(())
 }
 
+struct BonsaiClient {
+    pub(crate) client: Arc<Client>,
+}
+
+impl BonsaiClient {
+    pub fn instance(
+        encoded_image_id: &String,
+        elf: Vec<u8>,
+    ) -> Result<&'static BonsaiClient, BonsaiExecutionError> {
+        static INSTANCE: OnceCell<BonsaiClient> = OnceCell::new();
+
+        Ok(INSTANCE
+            .get_or_try_init(|| {
+                let client = Client::from_env(risc0_zkvm::VERSION)?;
+                client.upload_img(&encoded_image_id, elf)?;
+
+                Ok(BonsaiClient {
+                    client: Arc::new(client),
+                })
+            })
+            .map_err(|e| BonsaiExecutionError::SdkFailure(e))?)
+    }
```

Comment:
Personally, I believe that using a static shared instance can be risky. For instance, if the shared client instance becomes broken due to a network issue, is it possible for the shared client instance to automatically recover once the network problem is resolved?

How about checking whether the image has been uploaded when creating a new client instance? [`get_image_upload_url`](https://github.com/risc0/risc0/blob/main/bonsai/sdk/src/lib.rs#L569-L586)

```rust
            let res = self
                .client
                .get(format!("{}/images/upload/{}", self.url, image_id))
                .send()
                .await?;

            if res.status() == 204 {
                return Ok(ImageExistsOpt::Exists);
            }
```

Reply:
Oh, I notice that [`upload_img`](https://github.com/risc0/risc0/blob/main/bonsai/sdk/src/lib.rs#L608-L611) will not re-upload image if the image has been already uploaded. If so, what issue we are facing?

Reply:
I think your concerns is solid, here risc0 has no change, but sp1 side has a similar static instance, need to figure out a way to make sure the client is alive.